tfhe/src/integer/server_key/radix_parallel/div_mod.rs

```diff
@@ -0,0 +1,496 @@
+    /// let (q_res, r_res) = sks.div_rem_parallelized(&ct1, &ct2);
+    ///
+    /// // Decrypt:
+    /// let q: u64 = cks.decrypt(&q_res);
+    /// let r: u64 = cks.decrypt(&r_res);
+    /// assert_eq!(q, msg1 / msg2);
+    /// assert_eq!(r, msg1 % msg2);
+    /// ```
+    pub fn div_rem_parallelized(
+        &self,
+        numerator: &RadixCiphertext,
+        divisor: &RadixCiphertext,
+    ) -> (RadixCiphertext, RadixCiphertext) {
+        let mut tmp_numerator: RadixCiphertext;
+        let mut tmp_divisor: RadixCiphertext;
+
+        let (numerator, divisor) = match (
+            numerator.block_carries_are_empty(),
+            divisor.block_carries_are_empty(),
+        ) {
+            (true, true) => (numerator, divisor),
+            (true, false) => {
+                tmp_divisor = divisor.clone();
+                self.full_propagate_parallelized(&mut tmp_divisor);
+                (numerator, &tmp_divisor)
+            }
+            (false, true) => {
+                tmp_numerator = numerator.clone();
+                self.full_propagate_parallelized(&mut tmp_numerator);
+                (&tmp_numerator, divisor)
+            }
+            (false, false) => {
+                tmp_divisor = divisor.clone();
+                tmp_numerator = numerator.clone();
+                rayon::join(
+                    || self.full_propagate_parallelized(&mut tmp_numerator),
+                    || self.full_propagate_parallelized(&mut tmp_divisor),
+                );
+                (&tmp_numerator, &tmp_divisor)
+            }
+        };
+
+        self.unchecked_div_rem_parallelized(numerator, divisor)
+    }
+
+    pub fn smart_div_rem_parallelized(
+        &self,
+        numerator: &mut RadixCiphertext,
+        divisor: &mut RadixCiphertext,
+    ) -> (RadixCiphertext, RadixCiphertext) {
+        rayon::join(
+            || {
+                if !numerator.block_carries_are_empty() {
+                    self.full_propagate_parallelized(numerator)
+                }
+            },
+            || {
+                if !divisor.block_carries_are_empty() {
+                    self.full_propagate_parallelized(divisor)
+                }
+            },
+        );
+        self.unchecked_div_rem_parallelized(numerator, divisor)
+    }
+
+    //======================================================================
+    //                Div
+    //======================================================================
+
+    pub fn unchecked_div_assign_parallelized(
+        &self,
+        numerator: &mut RadixCiphertext,
+        divisor: &RadixCiphertext,
+    ) {
+        let (q, _r) = self.unchecked_div_rem_parallelized(numerator, divisor);
+        *numerator = q;
+    }
+
+    pub fn unchecked_div_parallelized(
+        &self,
+        numerator: &RadixCiphertext,
+        divisor: &RadixCiphertext,
+    ) -> RadixCiphertext {
+        let (q, _r) = self.unchecked_div_rem_parallelized(numerator, divisor);
```

Comment:
would it make sense at some point to ditch e.g. the remainder computation to go faster ? not saying we need to do it now but it could speed up just the div it feels like

Reply:
yeah, guessed as much

Reply:
thanks !